### PR TITLE
Use distroless image to build controller and speaker containers

### PR DESCRIPTION
**Pull request by minmzzhang**
Inspired by https://github.com/metallb/metallb/pull/2117
fixes https://github.com/metallb/metallb/issues/2110

A big thank you to @oribon and @georgettica for the detail discussion in #2117 which is **the** design document I followed to this solution. Also appreciate @oribon 's help during design implementation, that I was able to run local integration and e2e tests! 

**Is this a BUG FIX or a FEATURE ?**:

/kind feature

**What this PR does / why we need it**:
gcr.io/distroless/static base image is used to replace the original docker.io/alpine image, for a distroless solution that controller and speaker images will be built from scratch. Since distroless base image has no shell entry, there are two major impacts with this solution:

1. Some of the frr init containers which need to move files around will need a local copy tool. 
2. Some of the e2e tests which need to execute certain commands (e.g. "ifconfig") from inside the speaker will need to have an ephemeral container attached to the pod to complete the task. 

In this PR, a local copy tool (cp-tool) was introduced to cope with the first concern, and a podDebugExecutor was introduced under e2etest to attach a debug ephemeral container when there is a need to execute additional commands.

**Special notes for your reviewer**:

**Release note**:

```release-note
The speaker and the controller containers are now based on a distroless image, reducing the attack surface
```

### CODE REVIEW

File: e2etest/pkg/executor/executor.go [hunk 1]
```diff
@@ -68,3 +68,25 @@ func (p *podExecutor) Exec(cmd string, args ...string) (string, error) {
...
+		namespace: namespace,
+		name:      name,
+	}
+}
+
+func (pd *podDebugExecutor) Exec(cmd string, args ...string) (string, error) {
+	if Kubectl == "" {
+		return "", errors.New("the kubectl parameter is not set")
+	}
+	fullargs := append([]string{"debug", "-it", "-n", pd.namespace, pd.name, "-c", "pod-executor-debugger", "--image=busybox", "--", cmd}, args...)
```
Comment (oribon) [inline e2etest/pkg/executor/executor.go:89]: I think we should randomize the container name here `pod-executor-debugger-..`, the reason being that if you run the same debug container multiple times it only remembers the first output (the first debug command that caused the container to be terminated) which could cause confusion if the test is ran multiple times on the same cluster

File: e2etest/pkg/executor/executor.go [hunk 2]
```diff
@@ -68,3 +68,25 @@ func (p *podExecutor) Exec(cmd string, args ...string) (string, error) {
 	out, err := exec.Command(Kubectl, fullargs...).CombinedOutput()
 	return string(out), err
 }
+
+// add ephemeral container to deal with distroless image
+type podDebugExecutor struct {
+	namespace string
+	name      string
```
Comment (oribon) [inline e2etest/pkg/executor/executor.go:75]: wdyt about addibg an image for the container here (instead of defaulting to busybox)? this will make the call for `ForPodDebug` more resemblant to the usual `kubectl debug` where the caller specifies the image.

File: frr-tools/cp-tool/cp-tool.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+	if len(os.Args) != 3 {
+		level.Error(logger).Log("Please provide two command line arguments.")
+		return
+	}
+
+	sourceFile, destinationFile := os.Args[1], os.Args[2]
+	sourceFile = filepath.Clean(sourceFile)
+	destinationFile = filepath.Clean(destinationFile)
+
+	err = cp.Copy(sourceFile, destinationFile)
```
Comment (oribon) [inline frr-tools/cp-tool/cp-tool.go:32]: did you try something that doesn't require importing another dependency? I wonder if we can just do this with some `io.Copy` (or anything else from the standard library) that wouldn't be too messy since our usecase is pretty simple

File: frr-tools/cp-tool/cp-tool.go [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+	"path/filepath"
+
+	cp "github.com/otiai10/copy"
+
+	"github.com/go-kit/log/level"
+	"go.universe.tf/metallb/internal/logging"
+)
+
+func main() {
+	logger, err := logging.Init("error")
```
Comment (oribon) [inline frr-tools/cp-tool/cp-tool.go:17]: we don't really need a logger here, let's just panic / fmt.Printf on the few occasions where needed

**Review by oribon — COMMENTED** [review]
awesome work! thanks a lot
can you also squash your commits into one with the gist of the change?(https://metallb.io/community/#commit-messages)

File: e2etest/pkg/executor/executor.go (hunk 2 above)
Comment (minmzzhang) [reply e2etest/pkg/executor/executor.go:77]: added the randomized string for debugger.

File: e2etest/pkg/executor/executor.go (hunk 1 above)
Comment (minmzzhang) [reply e2etest/pkg/executor/executor.go]: A randomized string (length=5) was added to the pod executor debugger.

File: e2etest/pkg/executor/executor.go (hunk 2 above)
Comment (minmzzhang) [reply e2etest/pkg/executor/executor.go:77]: image name is configurable now.

File: frr-tools/cp-tool/cp-tool.go (hunk 3 above)
Comment (minmzzhang) [reply frr-tools/cp-tool/cp-tool.go]: Originally there was file permission issue for the local copy tool when using io.Copy, so I switched to a community package. With the new submission, I've added the preserve permission functionality, so it should be clean now.

File: frr-tools/cp-tool/cp-tool.go (hunk 4 above)
Comment (minmzzhang) [reply frr-tools/cp-tool/cp-tool.go]: logger removed.

File: frr-tools/cp-tool/cp-tool.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+func main() {
+	if len(os.Args) != 3 {
+		fmt.Printf("Please provide two command line arguments.")
+		return
+	}
+
+	sourceFile, destinationFile := os.Args[1], os.Args[2]
+	sourceFile = filepath.Clean(sourceFile)
+	destinationFile = filepath.Clean(destinationFile)
```
Comment (oribon) [inline frr-tools/cp-tool/cp-tool.go:70]: small nit: this is probably redundant with the one happening inside the function

File: frr-tools/cp-tool/cp-tool.go [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+import (
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+)
+
+func CopyFile(srcFile string, destFile string) error {
+	srcFile = filepath.Clean(srcFile)
+	destFile = filepath.Clean(destFile)
```
Comment (oribon) [inline frr-tools/cp-tool/cp-tool.go:14]: nit: rename the input to src,dst (so the new vars are declared with new names)

File: frr-tools/cp-tool/cp-tool.go [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+	defer src.Close()
+
+	srcStat, err := os.Stat(srcFile)
+	if err != nil {
+		err = fmt.Errorf("failed to check stats for %s: %w", srcFile, err)
+		return err
+	}
+
+	srcPerm := srcStat.Mode().Perm()
+	fmt.Printf("File permission before copying %v \n", srcPerm)
```
Comment (oribon) [inline frr-tools/cp-tool/cp-tool.go:29]: nit: I think we can remove it (unless you think this adds value)

File: frr-tools/cp-tool/cp-tool.go [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+	}
+
+	dstStat, err := os.Stat(destFile)
+	if err != nil {
+		err = fmt.Errorf("failed to check stats for %s: %w", destFile, err)
+		return err
+	}
+
+	dstPerm := dstStat.Mode().Perm()
+	fmt.Printf("File permission After copying %v \n", dstPerm)
```
Comment (oribon) [inline frr-tools/cp-tool/cp-tool.go:57]: same?

**Review by oribon — APPROVED** [review]
left a few small comments, looks really good!
also, please change the commit message to be more about "what is the change" and not "how we changed" :) 
approved, cc @fedepaol to also give this a look

File: frr-tools/cp-tool/cp-tool.go (hunk 5 above)
Comment (minmzzhang) [reply frr-tools/cp-tool/cp-tool.go]: removed the redundant part.

File: frr-tools/cp-tool/cp-tool.go (hunk 6 above)
Comment (minmzzhang) [reply frr-tools/cp-tool/cp-tool.go]: renamed the inputs

File: e2etest/pkg/executor/executor.go [hunk 9]
```diff
@@ -68,3 +70,34 @@ func (p *podExecutor) Exec(cmd string, args ...string) (string, error) {
...
+		image:     image,
+	}
+}
+
+func (pd *podDebugExecutor) Exec(cmd string, args ...string) (string, error) {
+	if Kubectl == "" {
+		return "", errors.New("the kubectl parameter is not set")
+	}
+
+	imagearg := "--image=" + pd.image
```
Comment (fedepaol) [inline e2etest/pkg/executor/executor.go:96]: nit: imageArg, targetArg, debuggerArg

File: e2etest/l2tests/interface_selector.go [hunk 10]
```diff
@@ -170,7 +170,7 @@ var _ = ginkgo.Describe("L2-interface selector", func() {
 			}, 1*time.Minute, 1*time.Second).Should(Not(HaveOccurred()))
 			speakerPod, err := metallb.SpeakerPodInNode(cs, svcNode.Name)
 			Expect(err).NotTo(HaveOccurred())
-			selectorMac, err := mac.GetIfaceMac(NodeNics[0], executor.ForPod(speakerPod.Namespace, speakerPod.Name, "speaker"))
+			selectorMac, err := mac.GetIfaceMac(NodeNics[0], executor.ForPodDebug(speakerPod.Namespace, speakerPod.Name, "speaker", "busybox"))
```
Comment (fedepaol) [inline e2etest/l2tests/interface_selector.go:173]: the only concern I have here is relying on another image (from dockerhub, also), which may result in hitting the pull limits. How about using the agnhost image we are already using [here](https://github.com/metallb/metallb/blob/main/e2etest/pkg/jigservice/jig.go#L42) (assuming it already has all we need to debug)?

File: website/content/troubleshooting/_index.md [hunk 11]
```diff
@@ -288,3 +288,11 @@ Additionally, the status of the service and of the endpoints must be provided:
 kubectl get endpointslices <my-service> -o yaml
 kubectl get svc <my_service> -o yaml
 ```
+
+### How to debug a container with distroless image
```
Comment (fedepaol) [inline website/content/troubleshooting/_index.md:292]:
```suggestion
### How to debug the speaker and the controller containers
```

File: website/content/troubleshooting/_index.md [hunk 12]
```diff
@@ -288,3 +288,11 @@ Additionally, the status of the service and of the endpoints must be provided:
...
+
+With the distroless solution, if someone needs to debug inside speaker container, an ephemeral container should be used:
```
Comment (fedepaol) [inline website/content/troubleshooting/_index.md:294]:
```suggestion
Due to the fact that both the speaker and the controller containers are based on a distroless image, an ephemeral container should be used to debug inside them: 
```

**Comment by fedepaol** [comment]
Also, I'd reword the release notes section, saying that now the speaker and the controller are based on distroless.

File: e2etest/pkg/executor/executor.go (hunk 9 above)
Comment (minmzzhang) [reply e2etest/pkg/executor/executor.go:96]: changed

File: e2etest/l2tests/interface_selector.go (hunk 10 above)
Comment (minmzzhang) [reply e2etest/l2tests/interface_selector.go:173]: good idea! changed.

File: website/content/troubleshooting/_index.md (hunk 12 above)
Comment (minmzzhang) [reply website/content/troubleshooting/_index.md:294]: modified.

File: website/content/troubleshooting/_index.md (hunk 11 above)
Comment (minmzzhang) [reply website/content/troubleshooting/_index.md:292]: changed..

**Comment by fedepaol** [comment]
LGTM
